llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -16075,56 +16075,50 @@ SITargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *RMW) const {
       return AtomicExpansionKind::CmpXChg;
     }
 
-    if (!Ty->isFloatTy() && (!Subtarget->hasGFX90AInsts() || !Ty->isDoubleTy()))
+    if (!AMDGPU::isFlatGlobalAddrSpace(AS) &&
+        AS != AMDGPUAS::BUFFER_FAT_POINTER)
       return AtomicExpansionKind::CmpXChg;
 
-    if ((AMDGPU::isFlatGlobalAddrSpace(AS) ||
-         AS == AMDGPUAS::BUFFER_FAT_POINTER) &&
-        Subtarget->hasAtomicFaddNoRtnInsts()) {
-      if (Subtarget->hasGFX940Insts())
-        return AtomicExpansionKind::None;
+    // TODO: gfx940 supports v2f16 and v2bf16
+    if (Subtarget->hasGFX940Insts() && (Ty->isFloatTy() || Ty->isDoubleTy()))
```

Comment:
I lost the track of this condition after many changes. Can this hasGFX940Insts() be spelled out in terms of a specific feature/target property or its combination? What about gfx11/12, we seem to expand even f32?

Reply:
Should that be something like:
```
if  (Subtarget->hasAgentScopeFineGrainedRemoteMemoryAtomics() &&
     Subtarget->hasMemoryAtomicFaddF32DenormalSupport()) {
  if (Subtarget->hasAtomicFaddRtnInsts() && Ty->isFloatTy())
    return AtomicExpansionKind::None;
  if (Subtarget->hasFlatBufferGlobalAtomicFaddF64Inst() && Ty->isDoubleTy())
    return AtomicExpansionKind::None;
}
```
Maybe also check addresspace global vs flat.

I believe that was the original intent of the gfx940 check, but I am not sure the meaning of unsafe is still the same.

Reply:
I have created https://github.com/llvm/llvm-project/pull/100217